### DISCUSSION TITLE
How to register custom dataset module?

### DISCUSSION THREAD

**Discussion by seon-creator** [open-mmlab/mmpose#2520] [Q&A] [answered]
Hello, I'm trying to register custom module to use custom model for enter datatype easliy.
But just make custom keypoint information file, and make a class python file is not work correctly.

I tried like this.
# mmpose/mmpose/datasets/datasets/body
in this root, I made coco_arm.py
and make a class like this

`@DATASETS.register_module()
class CocoArm(BaseCocoStyleDataset):
    METAINFO: dict = dict(from_file='configs/_base_/datasets/coco_arm.py') # for arm`

The coco_arm.py is made by me.

After then, I add import line and enter the class name in the __all__, like this.

`from .aic_dataset import AicDataset
from .coco_dataset import CocoDataset
from .coco_arm import CocoArm   # edit
from .crowdpose_dataset import CrowdPoseDataset
from .jhmdb_dataset import JhmdbDataset
from .mhp_dataset import MhpDataset
from .mpii_dataset import MpiiDataset
from .mpii_trb_dataset import MpiiTrbDataset
from .ochuman_dataset import OCHumanDataset
from .posetrack18_dataset import PoseTrack18Dataset
from .posetrack18_video_dataset import PoseTrack18VideoDataset

__all__ = [
    'CocoDataset','CocoDatasetArm', 'MpiiDataset', 'MpiiTrbDataset', 'AicDataset',
    'CrowdPoseDataset', 'OCHumanDataset', 'MhpDataset', 'PoseTrack18Dataset',
    'JhmdbDataset', 'PoseTrack18VideoDataset', 'CocoArm' # add
]`

After finishing, I train the model, but error occur.
#KeyError: 'CocoArm is not in the dataset registry. Please check whether the value of `CocoArm` is correct or it was registered as expected. More details can be found at https://mmengine.readthedocs.io/en/latest/advanced_tutorials/config.html#import-the-custom-module'

How to register custom module?

**Comment by Ben-Louis** [comment]
Could you please try running `from mmpose.datasets.datasets import CocoArm` and check if this new class can be imported successfully?

**Reply by seon-creator** [reply]
Thank you for reply !
I saw the code and there is no problem for import.
I tried it like this but same error occurs.

# Copyright (c) OpenMMLab. All rights reserved.
from .aic_dataset import AicDataset
from .coco_dataset import CocoDataset
from mmpose.datasets.datasets.body.coco_arm import CocoArm   # edit from .coco_arm to mmpose.datasets.datasets.body.coco_arm
from .crowdpose_dataset import CrowdPoseDataset
from .jhmdb_dataset import JhmdbDataset
from .mhp_dataset import MhpDataset
from .mpii_dataset import MpiiDataset
from .mpii_trb_dataset import MpiiTrbDataset
from .ochuman_dataset import OCHumanDataset
from .posetrack18_dataset import PoseTrack18Dataset
from .posetrack18_video_dataset import PoseTrack18VideoDataset

__all__ = [
    'CocoDataset','CocoDatasetArm', 'MpiiDataset', 'MpiiTrbDataset', 'AicDataset',
    'CrowdPoseDataset', 'OCHumanDataset', 'MhpDataset', 'PoseTrack18Dataset',
    'JhmdbDataset', 'PoseTrack18VideoDataset', 'CocoArm' # add
]

# error message
KeyError: 'CocoArm is not in the dataset registry. Please check whether the value of `CocoArm` is correct or it was registered as expected.

Is it correctly try for your answer? If not, should I have to build by something code?
![code](https://github.com/open-mmlab/mmpose/assets/118324988/ae340707-4a96-431a-8ae2-0ffc5fb8fd3f)

**Reply by Tau-J** [reply]
Could you share a screenshot of the `mmpose/datasets/datasets/body/coco_arm.py`?

**Reply by seon-creator** [reply]
This is code of mmpose/datasets/datasets/body/coco_arm.py

![Screenshot from 2023-07-07 13-12-03](https://github.com/open-mmlab/mmpose/assets/118324988/d517e209-9ce6-4146-996c-f765854653f5)

**Reply by Tau-J** [reply, marked as answer]
Could you print `pip list | grep mm`? You need to install mmpose via `mim install -e .`

**Reply by seon-creator** [reply]
I found the reason!
I enter the command pip list | grep mm
and mmpose version was 1.0.0
I reinstall mmpose in virtual environment and it was solved.
Thank you so much !!
